File: src/triangulation_node.cpp
// Author: Keenan Burnett
#include <ros/ros.h>
#include <cv_bridge/cv_bridge.h>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>
#include <fstream>
#include "triangulation/triangulation_node.h"
#include <opencv2/core.hpp>
#include <opencv/cv.hpp>


void TriangulationNode::callback(const sensor_msgs::ImageConstPtr& img,
    const nav_msgs::OdometryConstPtr& odom) {
    auto start = std::chrono::high_resolution_clock::now();
    cv_bridge::CvImagePtr cv_ptr = cv_bridge::toCvCopy(img, sensor_msgs::image_encodings::RGB8);
    cv::Mat img_save = cv_ptr->image;
    std::ofstream mycsv;
    mycsv.open(triang_file, std::ios::app);
    typedef Eigen::Matrix<float, 1, Eigen::Dynamic> rowVectorf;
    typedef Eigen::Matrix<double, 1, Eigen::Dynamic> rowVectord;
    if (collect_image == 1){
      ros::Rate r(image_rate);
      double filetime = ros::Time::now().toSec();
      std::string filename = folder + std::to_string(filetime) + ".png";
      cv::imwrite(filename, img_save);
      zeus_tf::get_odom_tf(*odom, Toi);
      zeus_tf::get_transform(tfBuffer, camera_frame, "imu_link", Tic);
      E = zeus_tf::get_inverse_tf(Toi * Tic);
      for (int i = 0; i < 4; i++) {
          for (int j = 0; j < 4; j++) {
            float temp;
            temp = P(i,j);
            P_d(i, j) = static_cast<double>(temp);
          }
      }
      P_tilda = P_d * E; // print to csv file
      P_tilda.transposeInPlace();
      Tio.transposeInPlace();
      Eigen::VectorXd P_vector(Eigen::Map<Eigen::VectorXd>(P_tilda.data(), P_tilda.cols()*P_tilda.rows()));
      Eigen::VectorXd T_vector(Eigen::Map<Eigen::VectorXd>(Tio.data(), Tio.cols()*Tio.rows()));
      rowVectord P_row = P_vector.transpose();
      rowVectord T_row = T_vector.transpose();
      Eigen::VectorXf Pose = Eigen::VectorXf::Constant(3,1,1);
      Pose << odom->pose.pose.position.x, odom->pose.pose.position.y, odom->pose.pose.position.z;
      rowVectorf Pose_row = Pose.transpose();
      mycsv << std::setprecision(16) << filename << ";" << P_row << ";" << T_row << ";" << Pose_row << ";" << "\n";
      mycsv.close();
      r.sleep();
    }
    auto stop = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> elapsed = stop - start;
    ROS_DEBUG_STREAM("[OBJ] TRIANGULATION TIME: " << elapsed.count());
}

void TriangulationNode::initialize_transforms() {
    tfBuffer = zeus_tf::tfBufferPtr(new zeus_tf::tfBuffer());
    tfListener = zeus_tf::tfListenerPtr(new zeus_tf::tfListener(*tfBuffer));
    zeus_tf::get_transform(tfBuffer, "imu_link", "velodyne", Tvi);
    zeus_tf::get_transform(tfBuffer, "velodyne", camera_frame, Tcv);
}

void TriangulationNode::get_ros_parameters() {
    nh_.getParam(node_name + "/camera_frame",            camera_frame);
    nh_.getParam(node_name + "/image_rate",              image_rate);
    nh_.getParam(node_name + "/triang_file",             triang_file);
    nh_.getParam(node_name + "/folder",                  folder);
    nh_.getParam(node_name + "/odom",                    odom);
    nh_.getParam(node_name + "/collect_image",           collect_image);
}

File: src/triangulation/triangulation_node.py
#! /usr/bin/python
import rospy
import cv2, csv
from cv_bridge import CvBridge
import numpy as np
import std_msgs, geometry_msgs, nav_msgs, sensor_msgs, tf2_geometry_msgs
import tf2_ros, tf
import os

class transforms:

    def get_inverse_tf(self, T):
        inv = np.identity(4)

        R = T[0:3,0:3]
        t = T[0:3,3]
        inv[0:3,0:3] = R.transpose()
        inv[0:3,3] = np.matmul(-R.transpose(), t)
        return inv

    def get_odom_tf(self, odom):
        init_x = 0.0
        init_y = 0.0
        init_z = 0.0
        init_pos_set = False
        T = self.get_odom_tf_ex(odom, init_x, init_y, init_z, init_pos_set, False)
        return T

    def get_odom_tf_ex(self, odom, init_x, init_y, init_z, init_pos_set, zero_odom):
        Toi = np.zeros(shape=(4,4))
        if zero_odom:
            if init_pos_set == False:
                init_x = odom.pose.pose.position.x
                init_y = odom.pose.pose.position.y
                init_z = odom.pose.pose.position.z
                init_pos_set = True
        EPS = 1e-15
        x = odom.pose.pose.position.x - init_x
        y = odom.pose.pose.position.y - init_y
        z = odom.pose.pose.position.z - init_z
        qx = odom.pose.pose.orientation.x
        qy = odom.pose.pose.orientation.y
        qz = odom.pose.pose.orientation.z
        qw = odom.pose.pose.orientation.w
        p = np.matrix([[x],[y],[z]])
        qbar = np.matrix([[qx],[qy],[qz],[qw]])
        R = np.identity(3)
        if np.matmul(qbar.transpose(),qbar) > EPS:
            epsilon = np.matrix([[qx],[qy],[qz]])
            eta = qw
            epsilon_cross = np.matrix([[0, -epsilon[2], epsilon[1]], [epsilon[2], 0, -epsilon[0]], [-epsilon[1], epsilon[0], 0]])
            I = np.identity(3)
            R = (eta**2 - np.matmul(epsilon.transpose(), epsilon)[0,0])*I + np.subtract(2*(epsilon*epsilon.transpose()), 2*eta*epsilon_cross)
        Toi[0:3,0:3] = R.transpose()
        Toi[0:3,3] = p.reshape(3)
        Toi[3,0:3] = np.zeros(shape=(1,3))
        Toi[3,3] = 1.0
        return Toi

    def convert_to_eigen(self, tf):
        q_ros = tf.transform.rotation
        p_ros = tf.transform.translation
        T = np.zeros(shape=(4,4))
        p = np.matrix([[p_ros.x], [p_ros.y], [p_ros.z]])
        epsilon = np.matrix([[q_ros.x], [q_ros.y], [q_ros.z]])
        eta = q_ros.w
        epsilon_cross = np.matrix([[0, -epsilon[2], epsilon[1]], [epsilon[2], 0, -epsilon[0]], [-epsilon[1], epsilon[0], 0]])
        I = np.identity(3)
        R = (eta**2 - np.matmul(epsilon.transpose(), epsilon)[0,0])*I + np.subtract(2*(epsilon*epsilon.transpose()), 2*eta*epsilon_cross)
        p = -1 * R * p
        T[0:3,0:3] = R.transpose()
        T[0:3,3] = p.reshape(3)
        T[3,0:3] = np.zeros(shape=(1,3))
        T[3,3] = 1.0
        return T

    def get_transform(self, src_frame, tgt_frame):
        got_tf = False
        timeout = 0.025
        tfBuffer = tf2_ros.Buffer()
        listener = tf2_ros.TransformListener(tfBuffer)
        while (not got_tf and not rospy.is_shutdown()):
            try:
                tf_temp = tfBuffer.lookup_transform(src_frame, tgt_frame, rospy.Time(), rospy.Duration(timeout))
                got_tf = True
            except(tf2_ros.LookupException, tf2_ros.ConnectivityException, tf2_ros.ExtrapolationException):
                got_tf = False
                rospy.logwarn_throttle(1, "Transform from " + src_frame + " to " + tgt_frame + " was not found! Is TF tree and clock being published?")
            if got_tf:
                T = self.convert_to_eigen(tf_temp)
            else:
                T = np.zeros(shape=(4,4))
        return T

class TriangulationNode:
    def set_camera_matrix(self, CAM):
        P_ = CAM
        return True

    def set_node_name(self):
        global node_name
        node_name = rospy.get_name()
        return True

    def initialize_transforms(self):
        tfBuffer = tf2_ros.Buffer()
        tfListener = tf2_ros.TransformListener(tfBuffer)
        global Tic
        x = transforms()
        Tic = x.get_transform("mono_link", "imu_link")
        return True

    def get_ros_parameters(self):
        global camera_frame
        global image_rate
        global triang_file
        global folder
        global collect_image
        camera_frame = rospy.get_param(node_name + "/camera_frame")
        image_rate = rospy.get_param(node_name + "/image_rate")
        triang_file = rospy.get_param(node_name + "/triang_file")
        folder = rospy.get_param(node_name + "/folder")
        collect_image = rospy.get_param(node_name + "/collect_image")
        return True

def callback(img, odom):
    bridge = CvBridge()
    start = rospy.rostime.Time.now()
    node_name = rospy.get_name()
    img_save = bridge.imgmsg_to_cv2(img, desired_encoding="passthrough")
    data = open(triang_file, 'a')
    mycsv = csv.writer(data)
    if collect_image == 1:
        rospy.Rate(image_rate)
        filetime = rospy.rostime.Time.now()
        filename = folder + str(filetime) + ".png"
        cv2.imwrite(filename, img_save)
        transform = transforms()
        Toi = transform.get_odom_tf(odom)
        Tio = transform.get_inverse_tf(Toi)
        Tic = transform.get_transform(camera_frame, "imu_link")
        E = np.zeros(shape =(4,4))
        E = transform.get_inverse_tf(Toi * Tic)
        P = rospy.get_param(node_name + "/P")
        P = np.matrix(P)
        P = P.reshape(4,4)
        P_tilda = np.matmul(P, E)
        P_tilda = np.asarray(P_tilda).reshape(-1)
        Tio = np.asarray(Tio).reshape(-1)
        Pose = np.array([odom.pose.pose.position.x, odom.pose.pose.position.y, odom.pose.pose.position.z])
        row = [filename, P_tilda, Tio, Pose]
        mycsv.writerow(row)
        data.close()
    stop = rospy.rostime.Time.now()
    elapsed = stop - start
    rospy.logdebug("[OBJ] TRIANGULATION TIME: " + str(elapsed))
    return True
